xep-0388.xml

```diff
@@ -195,71 +300,187 @@
+  <p>As described in the <link url="#feature">SASL mechanism negotiation</link> section, the server can inform the client of one or more upgrades that need to be performed for the current account. The client SHOULD always request any upgrade task it recognises. To do this, it includes the 'upgrade' attribute in its &lt;authenticate/> element naming the upgrade task it wants to perform, as specified in <link url="#initiation">Initiation</link>.</p>
+  <p>Upon successfully authenticating the client (including any secondary authentication steps required for the account), but before the final &lt;success/> would be sent, the server sends a &lt;continue/> element, which MUST contain a single task, matching whatever was selected by the client. The client then initiates this upgrade task by providing a corresponding &lt;next/> element providing the task name.</p>
+  <p>Upon receiving the &lt;next/> element for the upgrade, the server provides the elements and data needed for the client to calculate the requested data. The concrete elements and exchanges needed for the upgrade are specific to individual tasks. This document currently defines tasks for upgrading credentials between different members of the SCRAM family of mechanisms. Other tasks may be documented in future versions of this document, or in other documents.</p>
+  <p>Upgrade task names SHOULD have a prefix of "UPGR-" followed by the SASL mechanism name to upgrade to, and if multiple mechanisms differ only in their support for channel-binding (e.g. SCRAM's -PLUS variants), implementations MUST use only the names of variants without channel-binding for the task names, because mechanism upgrades are independent of any channel-binding. Finally, upgrade tasks MUST NOT transmit plaintext passwords (or any reversible encoding of them) if the SASL mechanism to upgrade allows this to be avoided.</p>
+  <example caption="SASL mechanism upgrade required"><![CDATA[
+<!-- The Server requests the client to perform the upgrade task for BLOOP2 -->
+<continue xmlns='urn:xmpp:sasl:2'>
+  <additional-data>
+    SSdtIGJvcmVkIG5vdy4=
+  </additional-data>
+  <tasks>
+    <task>UPGR-BLOOP2</task>
+  </tasks>
+  <text>This account requires an upgrade to BLOOP2 as requested by the client</text>
+</continue>
+
+<!-- The Client initiates the task -->
+<next xmlns='urn:xmpp:sasl:2' task='UPGR-BLOOP2'/>
+
+<!-- The Server provides the needed data -->
+<data xmlns='urn:xmpp:bloop2:example'>
+  <data>U28sIG5leHQgRk9TREVNIC0gMjAxOCwgdGhhdCBpcy4uLg==</data>
+  <description>BLOOP2 is cool!</description>
+</data>
+
+<!-- The Client now responds -->
+<upgrade xmlns='urn:xmpp:bloop2:example'>
+  <response>Li4uSSdsbCBidXkgYSBiZWVyIGZvciB0aGUgZmlyc3QgcGVyc29uIHdoby4uLg==</response>
+  <finalize>42</finalize>
+</upgrade>
```

Comment:
See above comments on switching from base64 to XML. Even a <task xmlns='urn:xmpp:sasl:2'/> wrapper around these would simplify a lot of things.